src/circuit.rs

```diff
@@ -120,6 +121,49 @@ pub struct Circuit {
     pub(crate) rcv: Value<ValueCommitTrapdoor>,
 }
 
+impl Circuit {
+    /// Constructs a `Circuit` from the following components:
+    /// - `merkle_path`: a merkle path witness of the `input_note`
+    /// - `input_note`: a note spent in scope of the action
+    /// - `output_note`: a note created in scope of the action
+    /// - `fvk`: full viewing key owning the `input_note`
+    /// - `alpha`: a scalar used for randomization of the action spend validating key
+    /// - `rcv`: trapdoor for the action value commitment
+    ///
+    /// Returns `None` if the `input_note` is not owned by the `fvk`.
+    pub fn from_action_context(
+        merkle_path: tree::MerklePath,
+        input_note: Note,
+        output_note: Note,
+        fvk: FullViewingKey,
+        alpha: pallas::Scalar,
+        rcv: ValueCommitTrapdoor,
+    ) -> Option<Circuit> {
+        let scope = fvk.scope_for_address(&input_note.recipient())?;
```

Comment:
I think that an API built around the existing builder components may be simpler to reason about.

The `SpendInfo` type is already correctly constructed from the components `input_note, merkle_path, fvk` in `Builder::add_spend`, so we could encapsulate this with a public `SpendInfo::new(..) -> Option<Self>` constructor.

The `RecipientInfo` type is definitely the wrong type to use here, because we already have a constructed note and presumably don't want to run encryption again.

The `ActionInfo` type could potentially be repurposed / refactored to share codepaths with this API. One option is that this API be changed to take a `SpendInfo`, and then the builder and this external API would share most of the circuit witness generation logic.

EDIT: After updating the suggestion to do this, I much prefer it. `ActionInfo::build` would then be altered to use `Circuit::from_action_context_unchecked`.

Reply:
So the conclusion is:
- publicize `SpendInfo`
- add `SpendInfo::new`
- change `from_action_context` type to take input context packed in `SpendInfo`
- add `Circuit::from_action_context_unchecked`
- use this constructor in `Action::build`